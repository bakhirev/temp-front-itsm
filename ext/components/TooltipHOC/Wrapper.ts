import styled from 'styled-components';

import type { ITooltipHOCProps } from './index';

interface IWrapperProps {
  display?: ITooltipHOCProps['display'];
}

export const Wrapper = styled.div<IWrapperProps>`
  display: ${({ display = 'inline-block' }) => display};
  position: relative;
`;
