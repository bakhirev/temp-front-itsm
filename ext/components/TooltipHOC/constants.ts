export const TOOLTIP_HEIGHT = 24;
export const TOOLTIP_HORIZONTAL_PADDING = 8;
export const TOOLTIP_MARGIN = 8;

export const POINTER_HEIGHT = 4;
export const POINTER_WIDTH = 8;

export const SAFE_SPACE = 16;
