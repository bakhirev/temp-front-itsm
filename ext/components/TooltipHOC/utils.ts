import { css } from 'styled-components';

import { getContainingBlockOffset } from '../common/utils';

import {
  POINTER_HEIGHT,
  POINTER_WIDTH,
  SAFE_SPACE,
  TOOLTIP_HEIGHT,
  TOOLTIP_MARGIN,
} from './constants';

import type { ITooltipHOCProps } from './';

export const findTooltipDirection = (wrapperRect: DOMRect, tooltipRect: DOMRect) => {
  const { width: tooltipWidth } = tooltipRect;
  const {
    bottom: wrapperBottom,
    left: wrapperLeft,
    right: wrapperRight,
    width: wrapperWidth,
  } = wrapperRect;

  const viewPortHeight = window.innerHeight;
  const viewPortWidth = window.innerWidth;

  const tooltipHorizontalAvailableSpace = (tooltipWidth - wrapperWidth) / 2 + SAFE_SPACE;

  if (viewPortWidth - wrapperRight < tooltipHorizontalAvailableSpace) {
    return 'left';
  }

  if (wrapperLeft < tooltipHorizontalAvailableSpace) {
    return 'right';
  }

  if (viewPortHeight < wrapperBottom + TOOLTIP_MARGIN + TOOLTIP_HEIGHT + SAFE_SPACE) {
    return 'top';
  }

  return 'bottom';
};

export const calculateTooltipPosition = (
  target: Element,
  tooltip: Element,
  direction: ITooltipHOCProps['direction'] = 'bottom'
) => {
  const { width: tooltipWidth, height: tooltipHeight } = tooltip.getBoundingClientRect();
  const { width: targetWidth, height: targetHeight } = target.getBoundingClientRect();

  const { targetX, targetY } = getTargetOffset(target, targetWidth, targetHeight);
  const defaultOffset = getDefaultOffset(targetWidth, targetHeight, tooltipWidth, tooltipHeight);
  const { parentTop, parentLeft } = getContainingBlockOffset(tooltip);

  return {
    posX: targetX + defaultOffset[direction].left - parentLeft,
    posY: targetY + defaultOffset[direction].top - parentTop,
  };
};

export const getPointerPositionStyle = (pointerDirection: ITooltipHOCProps['direction']) => {
  const widthAndHeightDevision = POINTER_WIDTH / POINTER_HEIGHT;

  switch (pointerDirection) {
    case 'left':
      return css`
        left: -${POINTER_HEIGHT + widthAndHeightDevision}px;
        top: calc(50% - (${POINTER_HEIGHT}px / 2));
        transform: rotate(-90deg);
      `;

    case 'right':
      return css`
        right: -${POINTER_HEIGHT + widthAndHeightDevision}px;
        top: calc(50% - (${POINTER_HEIGHT}px / 2));
        transform: rotate(90deg);
      `;

    case 'bottom':
      return css`
        bottom: -${POINTER_HEIGHT}px;
        left: calc(50% - (${POINTER_WIDTH}px / 2));
        transform: rotate(180deg);
      `;

    case 'top':
    default:
      return css`
        left: calc(50% - (${POINTER_WIDTH}px / 2));
        top: -${POINTER_HEIGHT}px;
      `;
  }
};

const getTargetOffset = (target: Element, targetWidth: number, targetHeight: number) => {
  const { top: targetTop, left: targetLeft } = target.getBoundingClientRect();
  return {
    targetX: targetLeft + targetWidth / 2,
    targetY: targetTop + targetHeight / 2,
  };
};

const getDefaultOffset = (
  targetWidth: number,
  targetHeight: number,
  tooltipWidth: number,
  tooltipHeight: number
) => {
  const top = {
    left: -(tooltipWidth / 2),
    top: -(targetHeight / 2 + tooltipHeight + TOOLTIP_MARGIN),
  };
  const bottom = {
    left: -(tooltipWidth / 2),
    top: targetHeight / 2 + TOOLTIP_MARGIN,
  };
  const left = {
    left: -(tooltipWidth + targetWidth / 2 + TOOLTIP_MARGIN),
    top: -(tooltipHeight / 2),
  };
  const right = {
    left: targetWidth / 2 + TOOLTIP_MARGIN,
    top: -(tooltipHeight / 2),
  };

  return { top, bottom, left, right };
};
