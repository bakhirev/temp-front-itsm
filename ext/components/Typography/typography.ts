const FONT_FAMILY = `'VTB Group UI', sans-serif`;

interface IFontSize {
  11: string;
  14: string;
  16: string;
  18: string;
  20: string;
  24: string;
  28: string;
  34: string;
  40: string;
  48: string;
  56: string;
}

const FONT_SIZE: IFontSize = {
  11: '11px',
  14: '14px',
  16: '16px',
  18: '18px',
  20: '20px',
  24: '24px',
  28: '28px',
  34: '34px',
  40: '40px',
  48: '48px',
  56: '56px',
};

interface IFontWeight {
  regular: number;
  medium: number;
  semibold: number;
  bold: number;
}

const FONT_WEIGHT: IFontWeight = {
  regular: 400,
  medium: 500,
  semibold: 600,
  bold: 700,
};

interface ILineHeight {
  16: string;
  20: string;
  24: string;
  28: string;
  32: string;
  36: string;
  44: string;
  52: string;
  60: string;
}

const LINE_HEIGHT: ILineHeight = {
  16: '16px',
  20: '20px',
  24: '24px',
  28: '28px',
  32: '32px',
  36: '36px',
  44: '44px',
  52: '52px',
  60: '60px',
};

export interface ITypography {
  fontFamily: string;
  fontSize: IFontSize;
  fontWeight: IFontWeight;
  lineHeight: ILineHeight;
}

export const TYPOGRAPHY: ITypography = {
  fontFamily: FONT_FAMILY,
  fontSize: FONT_SIZE,
  fontWeight: FONT_WEIGHT,
  lineHeight: LINE_HEIGHT,
};
