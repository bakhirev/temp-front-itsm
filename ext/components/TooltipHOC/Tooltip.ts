import styled, { css } from 'styled-components';

import { DEFAULT_THEME, Z_INDEXES } from '../common';

import { getPointerPositionStyle } from './utils';
import { POINTER_HEIGHT, TOOLTIP_HEIGHT, TOOLTIP_HORIZONTAL_PADDING } from './constants';

import type { ITooltipHOCProps } from './index';

interface ITooltipProps {
  light?: boolean;
  pointerDirection: ITooltipHOCProps['direction'];
  position: { posX: number; posY: number };
  showTooltip: boolean;
}

export const Tooltip = styled.div<ITooltipProps>`
  ${({ light, pointerDirection, position: { posX, posY }, showTooltip, theme }) => css`
    align-items: center;
    background: ${light ? theme.color.neutral.white : theme.color.neutral[80]};
    border-radius: ${theme.borderRadius};
    box-shadow: ${light ? theme.boxShadow[12] : 'none'};
    color: ${light ? theme.color.neutral[90] : theme.color.neutral.white};
    display: flex;
    height: ${TOOLTIP_HEIGHT}px;
    left: ${posX}px;
    padding: 0 ${TOOLTIP_HORIZONTAL_PADDING}px;
    position: fixed;
    top: ${posY}px;
    visibility: ${showTooltip ? 'visible' : 'hidden'};
    white-space: nowrap;
    z-index: ${Z_INDEXES.TOOLTIP};

    &::after {
      border-color: transparent transparent
        ${light ? theme.color.neutral.white : theme.color.neutral[80]} transparent;
      border-style: solid;
      border-width: 0 ${POINTER_HEIGHT}px ${POINTER_HEIGHT}px ${POINTER_HEIGHT}px;
      content: '';
      height: 0;
      position: absolute;
      width: 0;
      ${getPointerPositionStyle(pointerDirection)};
    }
  `};
`;

Tooltip.defaultProps = {
  theme: DEFAULT_THEME,
};
