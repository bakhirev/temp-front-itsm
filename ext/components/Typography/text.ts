import styled, { css } from 'styled-components';

import { TYPOGRAPHY } from './typography';
import type { ITypography } from './typography';

export interface ITextProps {
  color?: string;
}

export const getTextStyle = ({
  color,
  fontSize,
  fontWeight,
  lineHeight,
}: {
  color?: string;
  fontSize: keyof ITypography['fontSize'];
  fontWeight: keyof ITypography['fontWeight'];
  lineHeight: keyof ITypography['lineHeight'];
}) => css`
  color: ${color || 'inherit'};
  font-family: ${TYPOGRAPHY.fontFamily};
  font-size: ${TYPOGRAPHY.fontSize[fontSize]};
  font-weight: ${TYPOGRAPHY.fontWeight[fontWeight]};
  line-height: ${TYPOGRAPHY.lineHeight[lineHeight]};
`;

export const HL1 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 56, fontWeight: 'bold', lineHeight: 60 })};
`;

export const HL2 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 48, fontWeight: 'bold', lineHeight: 52 })};
`;

export const H1 = styled.h1<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 40, fontWeight: 'bold', lineHeight: 44 })};
`;

export const H2 = styled.h2<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 34, fontWeight: 'bold', lineHeight: 36 })};
`;

export const H3 = styled.h3<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 28, fontWeight: 'semibold', lineHeight: 32 })};
`;

export const H4 = styled.h4<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 24, fontWeight: 'semibold', lineHeight: 28 })};
`;

export const H5 = styled.h5<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 20, fontWeight: 'semibold', lineHeight: 24 })};
`;

export const H6 = styled.h6<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 18, fontWeight: 'semibold', lineHeight: 24 })};
`;

export const Subtitle1 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 18, fontWeight: 'regular', lineHeight: 28 })};
`;

export const Subtitle2 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 16, fontWeight: 'semibold', lineHeight: 20 })};
`;

export const Subtitle3 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 14, fontWeight: 'semibold', lineHeight: 20 })};
`;

export const Body1Long = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 16, fontWeight: 'regular', lineHeight: 24 })};
`;

export const Body1Short = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 16, fontWeight: 'regular', lineHeight: 20 })};
`;

export const Body2Long = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 14, fontWeight: 'regular', lineHeight: 20 })};
`;

export const Body2Short = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 14, fontWeight: 'regular', lineHeight: 16 })};
`;

export const Caption1 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 11, fontWeight: 'regular', lineHeight: 16 })};
`;

export const Button1 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 16, fontWeight: 'medium', lineHeight: 24 })};
`;

export const Button2 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 14, fontWeight: 'medium', lineHeight: 20 })};
`;

export const Button3 = styled.div<ITextProps>`
  ${({ color }) => getTextStyle({ color, fontSize: 11, fontWeight: 'medium', lineHeight: 16 })};
`;
