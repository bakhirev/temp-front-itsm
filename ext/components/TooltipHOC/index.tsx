import React, { Component, createRef } from 'react';
import ReactDOM from 'react-dom';
import type { ComponentType, RefObject } from 'react';

import { Caption1 } from '../Typography';
import { getScrollableParents } from '../common/utils';

import { Tooltip } from './Tooltip';
import { Wrapper } from './Wrapper';
import { calculateTooltipPosition, findTooltipDirection } from './utils';

type Direction = 'bottom' | 'left' | 'right' | 'top';
type DisplayType = 'block' | 'inline-block';

const POINTER_DIRECTION: { [key: string]: Direction } = {
  bottom: 'top',
  left: 'right',
  right: 'left',
  top: 'bottom',
};
const TOOLTIP_DELAY = 1500;

export interface ITooltipHOCProps {
  /** Имя класса для переопределения стилей */
  className?: string;
  /** Контейнер, в котором при необходимости можно отрисовать тултип через ReactDOM.createPortal */
  container?: Element | null;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Направление тултипа */
  direction?: Direction;
  /** Параметр display, согласно которому будет отрисован оборачиваемый в HOC компонент. По умолчанию inline-block */
  display?: DisplayType;
  /** Светлый вариант тултипа */
  light?: boolean;
  /** Ссылка на элемент, относительно которого будет позиционироваться тултип, если оборачиваемый в HOC компонент не подходит */
  target?: Element | null;
  /** Текст тултипа */
  tooltip: string;
  /** Отобразить тултип с задержкой в 1.5 секунды */
  withDelay?: boolean;
}

interface ITooltipHOCState {
  pointerDirection: Direction;
  posX: number;
  posY: number;
  showTooltip: boolean;
}

export const createTooltipHOC = <P extends Object>(WrappedComponent: ComponentType<P>) =>
  class TooltipHOC extends Component<P & ITooltipHOCProps, ITooltipHOCState> {
    tooltipRef: RefObject<HTMLDivElement>;
    wrapperRef: RefObject<HTMLDivElement>;
    showTooltipTimer = 0;
    scrollableParents?: Array<Element> = undefined;

    constructor(props: P & ITooltipHOCProps) {
      super(props);
      this.tooltipRef = createRef();
      this.wrapperRef = createRef();

      this.state = {
        pointerDirection: POINTER_DIRECTION['bottom'],
        posX: 0,
        posY: 0,
        showTooltip: false,
      };
    }

    hideTooltip = () => {
      this.setState({ showTooltip: false });
    };

    componentDidMount() {
      window.addEventListener('scroll', this.hideTooltip);
      window.addEventListener('resize', this.hideTooltip);
    }

    componentDidUpdate() {
      if (!this.scrollableParents && this.tooltipRef.current) {
        this.scrollableParents = getScrollableParents(this.tooltipRef.current);
        this.scrollableParents?.forEach((el) => el.addEventListener('scroll', this.hideTooltip));
      }
    }

    componentWillUnmount() {
      clearTimeout(this.showTooltipTimer);
      this.scrollableParents &&
        this.scrollableParents.forEach((el) => el.removeEventListener('scroll', this.hideTooltip));
      window.removeEventListener('scroll', this.hideTooltip);
      window.removeEventListener('resize', this.hideTooltip);
    }

    handleMouseEnter = () => {
      const { withDelay, direction, target } = this.props;

      const tooltipNode = this.tooltipRef.current;
      const wrapperNode: Element | null | undefined = target || this.wrapperRef.current;
      if (!tooltipNode || !wrapperNode) return;

      this.showTooltipTimer = window.setTimeout(
        () => {
          const tooltipRect = tooltipNode.getBoundingClientRect();
          const wrapperRect = wrapperNode.getBoundingClientRect();

          const tooltipDirection: Direction | undefined =
            direction || findTooltipDirection(wrapperRect, tooltipRect);
          if (!tooltipDirection) return;

          const pointerDirection = POINTER_DIRECTION[tooltipDirection];

          const position = calculateTooltipPosition(wrapperNode, tooltipNode, tooltipDirection);
          const { posX, posY } = position;

          this.setState({
            pointerDirection,
            posX,
            posY,
            showTooltip: true,
          });
        },
        withDelay ? TOOLTIP_DELAY : 0
      );
    };

    handleMouseLeave = () => {
      clearTimeout(this.showTooltipTimer);
      this.hideTooltip();
    };

    renderTooltip() {
      const { light, tooltip } = this.props;
      const { pointerDirection, posX, posY, showTooltip } = this.state;

      return (
        <Tooltip
          light={light}
          pointerDirection={pointerDirection}
          position={{ posX, posY }}
          ref={this.tooltipRef}
          showTooltip={showTooltip}
        >
          <Caption1>{tooltip}</Caption1>
        </Tooltip>
      );
    }

    render() {
      const { className, container, dataTestId, display } = this.props;
      const portalContainer: Element | null | undefined = container;

      return (
        <Wrapper
          className={className}
          data-test-id={dataTestId}
          display={display}
          ref={this.wrapperRef}
          onMouseEnter={this.handleMouseEnter}
          onMouseLeave={this.handleMouseLeave}
        >
          <WrappedComponent {...this.props} />
          {portalContainer
            ? ReactDOM.createPortal(this.renderTooltip(), portalContainer)
            : this.renderTooltip()}
        </Wrapper>
      );
    }
  };
