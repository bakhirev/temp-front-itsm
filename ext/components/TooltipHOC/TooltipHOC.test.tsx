import React, { ReactNode } from 'react';
import { mount, shallow } from 'enzyme';

import { Caption1 } from '../Typography';
import { createTooltipHOC } from '../TooltipHOC';
import type { ITooltipHOCProps } from '../TooltipHOC';

import { Tooltip } from './Tooltip';

jest.mock('../common', () => ({
  DEFAULT_THEME: {
    borderRadius: '',
    color: {
      neutral: {
        80: '',
        white: '',
      },
    },
  },
  Z_INDEXES: {
    TOOLTIP: 777,
  },
}));

jest.mock('../Typography', () => ({
  // eslint-disable-next-line react/display-name
  Caption1: ({ children }: { children: ReactNode }) => <div>{children}</div>,
}));

describe('TooltipHOC', () => {
  jest.useFakeTimers();

  beforeEach(() => {
    jest.clearAllTimers();
  });

  // jsdom defines innerWidth and innerHeight as 1024 x 768
  const viewPortWidth = window.innerWidth;
  const viewPortHeight = window.innerHeight;

  const wrapperHeight = 20;
  const wrapperLeft = 30;
  const wrapperWidth = 100;

  const tooltipWidth = 120;

  const WrappedComponent = () => <div />;
  const WrappedComponentWithTooltip = createTooltipHOC(WrappedComponent);

  const mountWrappedComponentWithTooltip = (
    direction?: ITooltipHOCProps['direction'],
    withDelay?: ITooltipHOCProps['withDelay']
  ) =>
    mount(<WrappedComponentWithTooltip direction={direction} tooltip="" withDelay={withDelay} />);

  const getTooltipPosition = (direction: ITooltipHOCProps['direction']) => {
    switch (direction) {
      case 'top':
        return {
          posX: 20,
          posY: -8,
        };

      case 'left':
        return {
          posX: -98,
          posY: 10,
        };

      case 'right':
        return {
          posX: 138,
          posY: 10,
        };

      case 'bottom':
      default:
        return {
          posX: 20,
          posY: 28,
        };
    }
  };

  const createGetBoundingClientRectMockFn = ({
    bottom = 0,
    height = 0,
    left = 0,
    right = 0,
    top = 0,
    width = 0,
  }: {
    bottom?: number;
    height?: number;
    left?: number;
    right?: number;
    top?: number;
    width?: number;
  }) => (): DOMRect => ({
    bottom,
    height,
    left,
    right,
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    toJSON: () => {},
    top,
    width,
    x: left,
    y: top,
  });

  it('should render component', () => {
    const wrapper = shallow(<WrappedComponentWithTooltip tooltip="" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render wrapped component', () => {
    const wrapper = shallow(<WrappedComponentWithTooltip tooltip="" />);
    expect(wrapper.find(WrappedComponent)).toHaveLength(1);
  });

  it('should pass all props to wrapped component', () => {
    const props = {
      bar: 'bar',
      foo: 'foo',
      tooltip: '',
    };
    const wrapper = shallow(<WrappedComponentWithTooltip {...props} />);
    const wrappedComponent = wrapper.find(WrappedComponent);
    expect(wrappedComponent.props()).toEqual(props);
  });

  it('should render tooltip with provided text using typography component', () => {
    const tooltipText = 'foo bar baz';
    const wrapper = shallow(<WrappedComponentWithTooltip tooltip={tooltipText} />);
    const styledComponent = wrapper.find(Caption1);
    expect(styledComponent.prop('children')).toEqual(tooltipText);
  });

  it("shouldn't show tooltip right after component has been rendered", () => {
    const wrapper = mountWrappedComponentWithTooltip();
    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('showTooltip')).toBe(false);
  });

  it('should show tooltip when mouse enters component', () => {
    const wrapper = mountWrappedComponentWithTooltip();

    wrapper.simulate('mouseEnter');
    jest.runAllTimers();
    wrapper.update();

    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('showTooltip')).toBe(true);
  });

  it('should hide tooltip when mouse leaves component', () => {
    const wrapper = mountWrappedComponentWithTooltip();

    wrapper.simulate('mouseEnter');
    jest.runAllTimers();
    wrapper.update();

    wrapper.simulate('mouseLeave');
    jest.runAllTimers();
    wrapper.update();

    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('showTooltip')).toBe(false);
  });

  it('should show tooltip with 1.5 seconds delay if prop "withDelay" is provided', () => {
    const wrapper = mountWrappedComponentWithTooltip(undefined, true);

    wrapper.simulate('mouseEnter');
    jest.advanceTimersByTime(1500);
    wrapper.update();

    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('showTooltip')).toBe(true);
  });

  it('should render tooltip in bottom direction by default if direction is not provided', () => {
    const wrapper = mountWrappedComponentWithTooltip();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      height: wrapperHeight,
      left: wrapperLeft,
      width: wrapperWidth,
    });

    const tooltipNode = wrapper.find(Tooltip).getDOMNode();
    tooltipNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: tooltipWidth,
    });

    wrapper.simulate('mouseEnter');
    jest.runAllTimers();
    wrapper.update();

    const tooltipPosition = getTooltipPosition('bottom');

    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('position')).toEqual(tooltipPosition);
  });

  it('should render tooltip in top direction if there is no space bottom and direction is not provided', () => {
    const wrapper = mountWrappedComponentWithTooltip();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      bottom: viewPortHeight,
      height: wrapperHeight,
      left: wrapperLeft,
      width: wrapperWidth,
    });

    const tooltipNode = wrapper.find(Tooltip).getDOMNode();
    tooltipNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: tooltipWidth,
    });

    wrapper.simulate('mouseEnter');
    jest.runAllTimers();
    wrapper.update();

    const tooltipPosition = getTooltipPosition('top');

    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('position')).toEqual(tooltipPosition);
  });

  it('should render tooltip in left direction if tooltip is wider than wrapped component, there is no space right and direction is not provided', () => {
    const wrapper = mountWrappedComponentWithTooltip();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      height: wrapperHeight,
      left: wrapperLeft,
      right: viewPortWidth,
      width: wrapperWidth,
    });

    const tooltipNode = wrapper.find(Tooltip).getDOMNode();
    tooltipNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: tooltipWidth,
    });

    wrapper.simulate('mouseEnter');
    jest.runAllTimers();
    wrapper.update();

    const tooltipPosition = getTooltipPosition('left');

    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('position')).toEqual(tooltipPosition);
  });

  it('should render tooltip in right direction if tooltip is wider than wrapped component, there is no space left and direction is not provided', () => {
    const wrapper = mountWrappedComponentWithTooltip();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      height: wrapperHeight,
      width: wrapperWidth,
    });

    const tooltipNode = wrapper.find(Tooltip).getDOMNode();
    tooltipNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: tooltipWidth,
    });

    wrapper.simulate('mouseEnter');
    jest.runAllTimers();
    wrapper.update();

    const tooltipPosition = { posX: 108, posY: 10 };

    const tooltip = wrapper.find(Tooltip);
    expect(tooltip.prop('position')).toEqual(tooltipPosition);
  });

  it('should render tooltip in provided direction', () => {
    const directions: ITooltipHOCProps['direction'][] = ['bottom', 'left', 'right', 'top'];
    directions.forEach((direction) => {
      const wrapper = mountWrappedComponentWithTooltip(direction);

      const wrapperNode = wrapper.getDOMNode();
      wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
        height: wrapperHeight,
        width: wrapperWidth,
        left: wrapperLeft,
      });

      const tooltipNode = wrapper.find(Tooltip).getDOMNode();
      tooltipNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
        width: tooltipWidth,
      });

      wrapper.simulate('mouseEnter');
      jest.runAllTimers();
      wrapper.update();

      const tooltipPosition = getTooltipPosition(direction);

      const tooltip = wrapper.find(Tooltip);
      expect(tooltip.prop('position')).toEqual(tooltipPosition);
    });
  });
});
