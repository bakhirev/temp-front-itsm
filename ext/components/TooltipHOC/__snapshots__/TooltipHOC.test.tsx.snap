// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipHOC should render component 1`] = `
<styled.div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <WrappedComponent
    tooltip=""
  />
  <styled.div
    pointerDirection="top"
    position={
      Object {
        "posX": 0,
        "posY": 0,
      }
    }
    showTooltip={false}
    theme={
      Object {
        "borderRadius": "",
        "color": Object {
          "neutral": Object {
            "80": "",
            "white": "",
          },
        },
      }
    }
  >
    <Caption1 />
  </styled.div>
</styled.div>
`;
